main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala

```diff
@@ -0,0 +1,716 @@
+        res0 <- BoolUnification.unify(t1, t2, renv).mapErr(toTypeError(_, prov))
+        // If bool unification has any constrs in output, then we know it failed so the subst is empty
+        res =
+          if (res0._2.isEmpty) {
+            ResolutionResult.newSubst(res0._1)
+          } else {
+            ResolutionResult.constraints(List(TypingConstraint.Equality(t1, t2, prov)), p1 || p2)
+          }
+      } yield res
+
+
+    case (Kind.RecordRow, Kind.RecordRow) =>
+      // first simplify the types to get rid of assocs if we can
+      for {
+        (t1, _) <- simplifyType(tpe1, renv, eqEnv, loc)
+        (t2, _) <- simplifyType(tpe2, renv, eqEnv, loc)
+        res0 <- RecordUnification.unifyRows(t1, t2, renv).mapErr(toTypeError(_, prov))
+        // If eff unification has any constrs in output, then we know it failed so the subst is empty
+        res = if (res0._2.isEmpty) ResolutionResult.newSubst(res0._1) else throw InternalCompilerException("can't handle complex record stuff", SourceLocation.Unknown)
+      } yield res
+
+    case (Kind.SchemaRow, Kind.SchemaRow) =>
+      // first simplify the types to get rid of assocs if we can
+      for {
+        (t1, _) <- simplifyType(tpe1, renv, eqEnv, loc)
+        (t2, _) <- simplifyType(tpe2, renv, eqEnv, loc)
+        res <- SchemaUnification.unifyRows(t1, t2, renv).mapErr(toTypeError(_, prov))
+      } yield ResolutionResult.newSubst(res)
+
+    case (Kind.CaseSet(sym1), Kind.CaseSet(sym2)) if sym1 == sym2 =>
+      for {
+        (t1, _) <- simplifyType(tpe1, renv, eqEnv, loc)
+        (t2, _) <- simplifyType(tpe2, renv, eqEnv, loc)
+        res <- CaseSetUnification.unify(t1, t2, renv, sym1.universe, sym1).mapErr(toTypeError(_, prov))
+      } yield ResolutionResult.newSubst(res)
+
+    case (k1, k2) if KindUnification.unifiesWith(k1, k2) => resolveEqualityStar(tpe1, tpe2, prov, renv, eqEnv, loc)
+
+    case _ => Err(toTypeError(UnificationError.MismatchedTypes(tpe1, tpe2), prov))
+  }
+
+
+  /**
+    * Resolves the equality between the two given types.
+    *
+    * The types must have kinds that do not have special unification rules.
+    *
+    * θ ⊩ᵤ τ₁ = τ₂ ⤳ u; r
+    */
+  private def resolveEqualityStar(tpe1: Type, tpe2: Type, prov: Provenance, renv: RigidityEnv, eqEnv: ListMap[Symbol.AssocTypeSym, Ast.AssocTypeDef], loc: SourceLocation)(implicit flix: Flix): Result[ResolutionResult, TypeError] = (tpe1, tpe2) match {
+    // varU
+    case (x: Type.Var, t) if renv.isFlexible(x.sym) && !t.typeVars.contains(x) =>
+      Result.Ok(ResolutionResult.newSubst(Substitution.singleton(x.sym, t)))
+
+    // varU
+    case (t, x: Type.Var) if renv.isFlexible(x.sym) && !t.typeVars.contains(x) =>
+      Result.Ok(ResolutionResult.newSubst(Substitution.singleton(x.sym, t)))
+
+
+    // reflU
+    case (Type.Cst(c1, _), Type.Cst(c2, _)) if c1 == c2 => Result.Ok(ResolutionResult.elimination)
+    case (x: Type.Var, y: Type.Var) if (x == y) => Result.Ok(ResolutionResult.elimination)
```

Comment:
I know this is reflexivity, but it would be more natural to have as the first rule. imho. I leave it to you.